drivers/ad7606/ad7606.c

```diff
@@ -0,0 +1,220 @@
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include "stdio.h"
+#include "stdint.h"
+#include "stdlib.h"
+#include "stdbool.h"
+#include "platform_drivers.h"
+#include "ad7606.h"
+
+static const struct ad7606_chip_info ad7606_chip_info_tbl[] = {
+	[ID_AD7605_4] = {
+		.num_channels = 4,
+		.has_oversampling = false,
+	},
+	[ID_AD7606_4] = {
+		.num_channels = 4,
+		.has_oversampling = true,
+	},
+	[ID_AD7606_6] = {
+		.num_channels = 6,
+		.has_oversampling = true,
+	},
+	[ID_AD7606_8] = {
+		.num_channels = 8,
+		.has_oversampling = true,
+	},
+};
+
+int32_t ad7606_spi_read_bulk(struct ad7606_dev *dev)
+{
+	uint8_t size;
+	int32_t ret;
+
+	size = ad7606_chip_info_tbl[dev->device_id].num_channels * 2;
+	ret = gpio_set_value(dev->gpio_convst, 0);
+	if (ret < 0)
+		return ret;
+
+	ret = gpio_set_value(dev->gpio_convst, 1);
+	if (ret < 0)
+		return ret;
+
+	return spi_write_and_read(dev->spi_desc, dev->data, size);
+}
+
+int32_t ad7606_spi_read_samples(struct ad7606_dev *dev,
+				uint8_t channel,
+				uint16_t *adc_data)
+{
+	int32_t ret;
+
+	ret = ad7606_spi_read_bulk(dev);
+	if (ret < 0)
+		return ret;
+
+	*adc_data = (dev->data[channel * 2] << 8 | dev->data[channel * 2 + 1]);
+
+	return 0;
+}
+
+int32_t ad7606_reset(struct ad7606_dev *dev)
+{
+	int32_t ret;
+
+	ret = gpio_set_value(dev->gpio_reset, 1);
+	if (ret < 0)
+		return ret;
+
+	mdelay(1);
+
+	return gpio_set_value(dev->gpio_reset, 0);
+}
+
+int32_t ad7606_request_gpios(struct ad7606_dev *dev,
+			     struct ad7606_init_param *init_param)
+{
+	int32_t ret;
+
+	ret = gpio_get(&dev->gpio_reset, init_param->gpio_reset);
+	if (ret < 0)
+		return ret;
+
+	ret = gpio_get(&dev->gpio_convst, init_param->gpio_convst);
+	if (ret < 0)
+		return ret;
+
+	ret = gpio_get(&dev->gpio_busy, init_param->gpio_busy);
+	if (ret < 0)
+		return ret;
+
+	ret = gpio_get(&dev->gpio_range, init_param->gpio_range);
+	if (ret < 0)
+		return ret;
+
+	if (!ad7606_chip_info_tbl[dev->device_id].has_oversampling)
+		return 0;
+
+	ret = gpio_get(&dev->gpio_range, init_param->gpio_os0);
+	if (ret < 0)
+		return ret;
+
+	ret = gpio_get(&dev->gpio_range, init_param->gpio_os1);
+	if (ret < 0)
+		return ret;
+
+	return gpio_get(&dev->gpio_range, init_param->gpio_os2);
+}
+
+int32_t ad7606_set_os_ratio(struct ad7606_dev *dev,
+			    enum ad7606_osr osr)
+{
+	int32_t ret;
+
+	ret = gpio_set_value(dev->gpio_os0, ((osr & 0x01) >> 0));
+	if (ret < 0)
+		return ret;
+
+	ret = gpio_set_value(dev->gpio_os1, ((osr & 0x02) >> 1));
+	if (ret < 0)
+		return ret;
+
+	return gpio_set_value(dev->gpio_os2, ((osr & 0x04) >> 2));
+}
+
+int32_t ad7606_init(struct ad7606_dev **device,
+		    struct ad7606_init_param *init_param)
+{
+	struct ad7606_dev *dev;
+	int32_t ret;
+
+	dev = (struct ad7606_dev *)malloc(sizeof(*dev));
```

Comment:
most drivers use `malloc()` in no-OS;
i'd argue that we could use `calloc()` to make sure that memory is 0

Reply:
fine with me